flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java

```diff
@@ -165,6 +168,69 @@ public void testDispatcherRecoversAfterLosingAndRegainingLeadership() throws Exc
 
     @Test
     public void testDirtyJobResultRecoveryInApplicationMode() throws Exception {
+        // having a dirty entry in the JobResultStore should make the ApplicationDispatcherBootstrap
+        // implementation fail to submit the job
+        final JobResultStore jobResultStore = new EmbeddedJobResultStore();
+        jobResultStore.createDirtyResult(
+                new JobResultEntry(
+                        TestingJobResultStore.createSuccessfulJobResult(
+                                ApplicationDispatcherBootstrap.ZERO_JOB_ID)));
+
+        final CompletableFuture<Throwable> fatalError = new CompletableFuture<>();
+        testJobResultStoreEntryLeadingToDuplicateJobSubmissionException(
+                jobResultStore,
+                ErrorHandlingSubmissionJob.createPackagedProgram(),
+                fatalError::complete);
+
+        FlinkAssertions.assertThatChainOfCauses(ErrorHandlingSubmissionJob.getSubmissionException())
```

Comment:
The fact that we have to rely on this singleton is a cause for concern. There should be a better way to retrieve the job result when using a mini cluster with application mode.